src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,225 @@
+        {
+            _originalUserPath = directory;
+            _originalFullPath = Path.GetFullPath(directory);
+            _recursive = recursive;
+            _predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+            _transform = transform ?? throw new ArgumentNullException(nameof(transform));
+            _threadId = Environment.CurrentManagedThreadId;
+        }
+
+        private FindEnumerable(
+            string originalUserPath,
+            string originalFullPath,
+            FindTransform<TResult> transform,
+            FindPredicate predicate,
+            bool recursive)
+        {
+            _originalUserPath = originalUserPath;
+            _originalFullPath = originalFullPath;
+            _predicate = predicate;
+            _transform = transform;
+            _recursive = recursive;
+            _threadId = Environment.CurrentManagedThreadId;
+        }
+
+        /// <summary>
+        /// Simple wrapper to allow creating a file handle for an existing directory.
+        /// </summary>
+        public static IntPtr CreateDirectoryHandle(string path)
+        {
+            IntPtr handle = Interop.Kernel32.CreateFile_IntPtr(
+                path,
+                Interop.Kernel32.FileOperations.FILE_LIST_DIRECTORY,
+                FileShare.ReadWrite | FileShare.Delete,
+                FileMode.Open,
+                Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle == IntPtr.Zero || handle == (IntPtr)(-1))
+            {
+                // Historically we throw directory not found rather than file not found
+                int error = Marshal.GetLastWin32Error();
+                if (error == Interop.Errors.ERROR_FILE_NOT_FOUND)
+                    error = Interop.Errors.ERROR_PATH_NOT_FOUND;
+
+                throw Win32Marshal.GetExceptionForWin32Error(error, path);
+            }
+
+            return handle;
+        }
+
+        public IEnumerator<TResult> GetEnumerator()
+        {
+            if (Interlocked.Exchange(ref _state, 1) == 0 && _threadId == Environment.CurrentManagedThreadId)
+            {
+                InitEnumeration();
+                return this;
+            }
+
+            FindEnumerable<TResult> clone = new FindEnumerable<TResult>(_originalUserPath, _originalFullPath, _transform, _predicate, _recursive);
+            clone.InitEnumeration();
+            return clone;
+        }
+
+        private void InitEnumeration()
+        {
+            // We want to delay creating any handles until the enumerator is actually created to
+            // avoid keeping handles open any longer than needed. Once we actually start enumerating
+            // it's fair game to stash handles until we've finished processing the given directory.
+
+            _directoryHandle = CreateDirectoryHandle(_originalFullPath);
+            _currentPath = _originalFullPath;
+            _buffer = ArrayPool<byte>.Shared.Rent(4096);
```

Comment:
4096 just being a nice number that's much bigger than any of the structs that windows will stuff in it? I see wwe don't handle buffer being too small as it should never happen.

Reply:
Yes, it should never happen. 4K is a pretty common buffer size, and happens to be what FindFirstFile uses. We get multiple results back in the buffer. Note that I did try playing around with the size of the buffer and didn't see any advantage to going larger than this.